### PR TITLE
[Feat] Load tiff files and nd2

### PR DESCRIPTION
**Pull request by kmilo9999**
Added code to support tiff and nd2 files

### CODE REVIEW

File: jaworski_cellseg/load_data_widget.py [hunk 1]
```diff
@@ -2,19 +2,83 @@
...
-        image = BioImage(file_path)
-        bio_data = image.get_image_data("CZYX")
-        napari_viewer.add_image(
-            bio_data,
-            channel_axis=0,
-            name=[f"Channel {i}" for i in range(bio_data.shape[0])],
-        )
-        return bio_data, image.metadata
+        file_extension = file_path.suffix.lower()
+        if file_extension == ".tiff" or file_extension == ".tif":
```
Comment (cpaniaguam) [inline jaworski_cellseg/load_data_widget.py:14]:
```suggestion
        if file_extension in [".tiff", ".tif"]:
```

File: jaworski_cellseg/load_data_widget.py [hunk 2]
```diff
@@ -2,19 +2,83 @@
...
+                # Extract resolution-related tags
+                x_resolution = page.tags.get("XResolution", None)
+                y_resolution = page.tags.get("YResolution", None)
+
+                unit_name = None
+                z_physical_size = 1.0
+                if tiff.imagej_metadata:
+                    unit_name = tiff.imagej_metadata["unit"]
+                    if "spacing" in tiff.imagej_metadata:
+                        z_physical_size = tiff.imagej_metadata["spacing"]
```
Comment (cpaniaguam) [inline jaworski_cellseg/load_data_widget.py:35]: Use the get method using those as defaults.
```suggestion
                if tiff.imagej_metadata:
                    unit_name = tiff.imagej_metadata.get("unit")
                    z_physical_size = tiff.imagej_metadata.get("spacing", 1.0)
                else:
                    unit_name = None
                    z_physical_size = 1.0
```

File: jaworski_cellseg/load_data_widget.py [hunk 3]
```diff
@@ -2,19 +2,83 @@
...
+import tifffile
+import numpy as np
 
 
-def jwslab_load_bio_data(napari_viewer: napari.Viewer, file_path: Path):
-    print("enter jwslab_pre_process_file")
+def jwslab_load_bio_data(
+    napari_viewer: napari.Viewer, file_path: Path, physical_sizes: dict
+):
     try:
```
Comment (cpaniaguam) [inline jaworski_cellseg/load_data_widget.py:12]: It's best practice to include the error prone code only in `try` blocks (easier to debug later). It seems like what might go wrong is that a bad file path was provided. Consider creating two helper functions, one for tiff-like files and another for `[".oir", ".nd2"]` and handle the exceptions within them.

File: jaworski_cellseg/load_data_widget.py [hunk 4]
```diff
@@ -2,19 +2,83 @@
...
+
+                unit_name = None
+                z_physical_size = 1.0
+                if tiff.imagej_metadata:
+                    unit_name = tiff.imagej_metadata["unit"]
+                    if "spacing" in tiff.imagej_metadata:
+                        z_physical_size = tiff.imagej_metadata["spacing"]
+
+                print(x_resolution)
+                print(y_resolution)
```
Comment (cpaniaguam) [inline jaworski_cellseg/load_data_widget.py:38]: Consider using a logger.

File: jaworski_cellseg/load_data_widget.py [hunk 5]
```diff
@@ -2,19 +2,83 @@
...
+
+            # Assume the same unit is used on every axis
+            xyz["unit"] = getattr(metadata.pixels, "physical_size_x_unit")
+            physical_sizes.update(xyz)
+
+        else:
+            # Raise exception for unsupported formats
+            raise ValueError(f"Unsupported file format: {file_extension}")
     except Exception as e:
         print(e)
```
Comment (cpaniaguam) [inline jaworski_cellseg/load_data_widget.py:83]: If using a logger you could do this instead:
`logging.error("Error processing bioimage file.", exc_info=True)`

File: jaworski_cellseg/load_data_widget.py [hunk 6]
```diff
@@ -2,19 +2,83 @@
...
+            )
+            ## get physical size of the data
+            metadata = image.metadata.images[0]
+            xyz: dict = {
+                key: getattr(metadata.pixels, f"physical_size_{key}")
+                for key in ["x", "y", "z"]
+            }
+
+            # Assume the same unit is used on every axis
+            xyz["unit"] = getattr(metadata.pixels, "physical_size_x_unit")
```
Comment (cpaniaguam) [inline jaworski_cellseg/load_data_widget.py:76]:
```suggestion
            xyz["unit"] = xyz["x"]
```

File: jaworski_cellseg/load_data_widget.py [hunk 7]
```diff
@@ -2,19 +2,83 @@
...
+            physical_sizes.update(xyz)
+
+        else:
+            # Raise exception for unsupported formats
+            raise ValueError(f"Unsupported file format: {file_extension}")
```
Comment (cpaniaguam) [inline jaworski_cellseg/load_data_widget.py:81]: Consider using the [guard pattern](https://en.wikipedia.org/wiki/Guard_(computer_science)) and check this condition at the start. Makes for simpler and more readable code.

**Review by cpaniaguam — CHANGES_REQUESTED** [review]
Thanks for this! Suggestions include using a guard and a logger and other minor improvements.

File: jaworski_cellseg/load_data_widget.py (hunk 6 above)
Comment (kmilo9999) [reply jaworski_cellseg/load_data_widget.py:76]: This one doesnt apply

File: jaworski_cellseg/load_data_widget.py (hunk 5 above)
Comment (kmilo9999) [reply jaworski_cellseg/load_data_widget.py:83]: I created an issue to add logger

File: jaworski_cellseg/load_data_widget.py (hunk 7 above)
Comment (kmilo9999) [reply jaworski_cellseg/load_data_widget.py:81]: I refactorized the code in that block.

File: jaworski_cellseg/load_data_widget.py (hunk 1 above)
Comment (kmilo9999) [reply jaworski_cellseg/load_data_widget.py]: Done

File: jaworski_cellseg/load_data_widget.py (hunk 2 above)
Comment (kmilo9999) [reply jaworski_cellseg/load_data_widget.py]: Done

File: jaworski_cellseg/load_data_widget.py (hunk 3 above)
Comment (kmilo9999) [reply jaworski_cellseg/load_data_widget.py]: Done

File: jaworski_cellseg/load_data_widget.py (hunk 5 above)
Comment (kmilo9999) [reply jaworski_cellseg/load_data_widget.py]: #13

**Review by cpaniaguam — APPROVED** [review]
LGTM

**Pull request merged**